### PR TITLE
[REVIEW] Add chebyshev, canberra, minkowksi and hellinger distance metrics

### PR DESCRIPTION
**Pull request by mdoijade**
This change will be followed up by cuML side change to use these metrics and update the pytest accordingly.

### CODE REVIEW

**Comment by mdoijade** [comment]
@teju85 @cjnolet  please help with review.

**Review by cjnolet — CHANGES_REQUESTED** [review]
Looks great. Just very minor things.

File: cpp/include/raft/distance/chebyshev.cuh [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      x, y, m, n, k, lda, ldb, ldd, dOutput, fin_op, stream);
+  }
+}
+
+/**
+ * @brief the chebyshev distance matrix calculation
+ *  It computes the following equation: cij = max(cij, op(ai-bj))
+ * @tparam InType input data-type (for A and B matrices)
+ * @tparam AccType accumulation data-type
+ * @tparam OutType output data-type (for C and D matrices)
```
[the remark this answers is not in the document]
Comment (mdoijade) [reply cpp/include/raft/distance/chebyshev.cuh:124]: done.

File: cpp/include/raft/distance/chebyshev.cuh [hunk 2]
```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2018-2021, NVIDIA CORPORATION.
```
[the remark this answers is not in the document]
Comment (mdoijade) [reply cpp/include/raft/distance/chebyshev.cuh]: done.

**Comment by mdoijade** [comment]
@cjnolet @teju85 the cuML side PR is passing - https://github.com/rapidsai/cuml/pull/3990
the failure in those tests is in dbscan due to some recent numpy update. 
All tests pertaining to these distance metrics are passing.

**Comment by cjnolet** [comment]
@mdoijade, the dbscan / numpy issue should be fixed in this PR: https://github.com/rapidsai/cuml/pull/4012. 

Would it reduce the compile time at all if we just used function arguments for the individual distances and removed the explicit templates for them?

**Review by cjnolet — APPROVED** [review]
LGTM

**Review by dantegd — CHANGES_REQUESTED** [review]
Before we merge this PR, could you quantify the impact of the changes in cuML's compilation time (locally at least)?

**Comment by mdoijade** [comment]
> 
> 
> Before we merge this PR, could you quantify the impact of the changes in cuML's compilation time (locally at least)?

@dantegd 
Without new distance metrics cuML total build time
real    9m24.016s
user    118m8.600s
sys     3m43.230s

With these 4 additional distance metrics cuML total build time
real    18m11.113s
user    120m28.920s
sys     3m34.880s

It is roughly 2x build time increase, and this reduction is after my change of using compiled cuML pairwise dist kernels instead of each time calling raft API wherever it is feasible without that change build time is about 25mins.
This increase is coming from pairwise_distance.cu

**Comment by mdoijade** [comment]
@dantegd @cjnolet @teju85 
I split the `pairwise_distance.cu` compilation into multiple source files for each distance metric to enable parallel build in it,
this has helped by a certain extent from previous effort like around 2-3 mins build time reduction.
I am seeing that minkowksi kernels are taking longer than others with CUDA 11.0 it has "powf(), pow()"  functions in it that is the only difference.
so after disabling minkowski kernels (other 3 new kernels enabled) the time take for full cuML build with CUDA 11.0
real    9m35.232s
user    113m56.715s
sys     3m44.966s
this is almost same time as without these kernels.

With these 4 additional distance metrics cuML total build time without splitting - CUDA 11.0 
real 18m11.113s
user 120m28.920s
sys 3m34.880s

Without new distance metrics cuML build time - CUDA 11.0 (current baseline)
real    9m24.016s
user    118m8.600s
sys     3m43.230s


After splitting the distance prims in `pairwise_distance.cu` into individual source files (with minkowski) - CUDA 11.0
real    17m44.648s
user    127m42.492s
sys     3m48.588s

Without minkowski  - CUDA 11.0
real    9m35.232s
user    113m56.715s
sys     3m44.966s


With minkowski  - cuda 11.2 + split pairwise distance 
real    9m57.291s
user    123m47.790s
sys     3m54.819s

With minkowski  - cuda 11.2 + without split pairwise distance 
real    12m22.500s
user    118m22.608s
sys     3m45.721s

So it is evident that this splitting of `pairwise_distance.cu` does help reduce build time.
and there is some compilation slowness issue with cuda 11.0 for `pow(), powf()` which seems to be resolved in cuda 11.2.

**Comment by mdoijade** [comment]
@dantegd can we merge this now? and also cuML PR - https://github.com/rapidsai/cuml/pull/3990

**Review by dantegd — APPROVED** [review]

**Comment by dantegd** [comment]
@gpucibot merge

**Pull request merged**